### ISSUE TITLE
Optimiser: simplification rule for rounding

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#3042]
> In some locations we use (x + 31) / 32 * 32 for rounding, but (x + 31) & ~31 should be a bit more efficient because it has only one constant.
> Also add that rule to the optimizer.

(From https://www.pivotaltracker.com/story/show/108204956)

**Comment by axic** [comment]
@ekpyron @chriseth wasn't this added?

**Comment by chriseth** [comment]
I don't think so.

**Comment by axic** [comment]
> In some locations we use (x + 31) / 32 * 32 for rounding, but (x + 31) & ~31 should be a bit more efficient because it has only one constant.

`mul(div(add(x, 31), 32), 32)` -> `and(add(x, 31), not(31))`

Or are we thinking about a generic rule:
`mul(div(add(x, a), b), b)` -> `and(add(x, a), not(a))` where `b == a + 1`, `b >= 1`, and `b` is a power of 2?

**Comment by chriseth** [comment]
could also be more generic

**Comment by Czar102** [comment]
From what I see in the bytecode output of Solidity 0.8.31, this is already implemented.

**Comment by cameel** [comment]
You're right. I just checked and `ExpressionSimplifier` handles this, which means that our rule list already covers it. I think this is because the rule we need is a bit simpler - just `MUL`/`DIV` by the same constant. Since the constant is a power of 2, these get converted to `SHL`/`SHR` and we already have a rule covering this pair: https://github.com/argotorg/solidity/blob/64118f21280f0196f491689f0fc93dc5bec20dc1/libevmasm/RuleList.h#L513-L528

So we did not really implement it, but it's not even necessary with other rules in place.

The caveat is of course that this is only for IR, but we would not be implementing it for evmasm as a part of this issue anyway. It will also not work on pre-constantinople EVMs as those lacked bit shifts, but again, [support for those is deprecated](https://github.com/argotorg/solidity/issues/12246#issuecomment-3848357648) and we don't promise that all optimizations will work with older EVM versions anyway.

**Comment by cameel** [comment]
This means that #6015 is done as well. I removed both from [Assorted Optimizer Ideas](https://notes.argot.org/@solidity-assorted-optimizer-ideas).
